tree.rb

```diff
@@ -0,0 +1,123 @@
+# ! MINOR BUGS: 
+# 1) :text is being deleted from the <ul> node. I know this has to do with 
+# the special gsub for <li> children but not specifically why it's happening.
+# 2) Spacing error on some of the other :text fields. Specifically the 'outer-div'
+
+# Create a DOM Tree Class
+
+# First thing is to create the Node struct
+Node = Struct.new(:name, :text, :classes, :id, :children, :parent)
+
+class DOMTree
+	attr_reader :file
+	attr_accessor :head
+
+	def initialize(file_name)
+		# gsub(/<!.*?>/, "") Removes <! doctype > from the document
+		# gsub(/>\s*</,"><") Removes all the whitespace from between > & < characters
+		# gsub(/\n/, "") Removes the newline characters from the document
+		@file = load(file_name).read.gsub(/\n/, "").gsub(/>\s*</,"><").gsub(/<!.*?>/, "")
+		@head = Node.new("Document", @file)
+	end
+
+	def load(file_name)
+		File.open(file_name, "r")
+	end
+
+	# Recursively get the children for each node. Returns nil
+	# if there are no more children.
+	def create_tree
+		queue = [@head]
+		while !queue.empty?
+			current_node = queue.pop
+			if has_children?(current_node.text)
+				current_node.children = []
+				create_children(current_node, queue)
+			else
+				current_node.text = current_node.text.strip
+			end
+		end
+	end
+
+	def create_children(node, queue)
+		while has_children?(node.text)
+			first_tag = get_first_tag(node.text)
+			child_name = get_name(first_tag)
+			child_class = (get_class(first_tag) == false ? nil :  get_class(first_tag))
+			child_id = (get_id(first_tag) == false ? nil :  get_id(first_tag))
+			child_text = (child_name == "li" ? get_inner_HTML(node.text, child_name, true) : get_inner_HTML(node.text, child_name))
+			child_node = Node.new(child_name, child_text, child_class, child_id, nil, node)
+			node.children << child_node
+			queue.unshift(child_node)
+			if child_name == "li" # Need to make things lazy for <li>
+				node.text = node.text.gsub(/^.*?(<#{child_name}.*?>.*?<\/#{child_name}>)/, "")
+			else
+				node.text = node.text.gsub(/(<#{child_name}.*?>.*<\/#{child_name}>)/, "")
+			end
+		end
+	end
+
+	# Return the HTML from between a pair of tags by name.
+	def get_inner_HTML(input, name, lazy = false)
+		if lazy == true
+			matches = input.match(/^.*?<#{name}.*?>(.*?)<\/#{name}>/)
+		else
+			matches = input.match(/<#{name}.*?>(.*)<\/#{name}>/)
+		end
+		matches ? matches.captures[0] : false
+	end
+
+	def has_children?(input)
+		# Looking for any kind of <> in the text. Remember, <...> tags might not
+		# be at the beginning.
+		matches = input.match(/<(.*?)>/)
+		(get_first_tag(input) == false && matches.nil?) ? false : true 
```

Comment:
Don't need a ternary to have a simple `false`/`true` output, could just use a bang `!` on the original expression